### ISSUE TITLE
Regex based generic file format

### ISSUE THREAD

**Issue by andrewleech** [WeblateOrg/weblate#2978]
**My problem**
I'm starting to use weblate to improve translating in an open source android app I maintain.
It works great for the standard strings (thanks!) but there's also a couple of other kinds of files I'd like to support:
* `string-xyz.txt`; A number of monolingual plain text files where each line is a separate sentence requiring translation
* Android string arrays; I've seen these are a known issue with a reference based workaround, however there are a lot of arrays in the app and breaking them all out will be unpleasant and harder to maintain.

**Describe the solution you'd like**
A file format handler that takes a custom regex string to parse out each of the strings. 
Something like this could be used to cover both these, and many other file formats.

My idea comes from the regex matcher already used by the Component discovery addon, but for file contents rather than file names. It could have a couple of named regex groups to specify context (if it exists) and source string. Basically each match is a new string.

If context doesn't exist in the file in question it means each language file would need to match the source structure exactly, but I guess this would be the case for any such file format anyway.

**Describe alternatives you've considered**
Alternatively, I convert all my plain text files to json, then build a new file reader in the app to handle the json files instead of the plain text files.

**Comment by nijel** [comment]
Regexp might work for plain text files, but I think it's bad idea for XML based formats such as Android.

Anyway for Android I think it's better to implement native support, see https://github.com/translate/translate/issues/3225

**Comment by andrewleech** [comment]
Yes, I agree that native support would be better in this case, but 5 years later it hasn't really progressed... I didn't even find that issue yesterday when I was searching through their tracker.

With regex it likely isn't possible to capture the Android array name as context for all the items in the array (unless the regex supported repeated groups of target strings in a single pattern), but it would be better than nothing.

While regex isn't the ideal, it is very flexible so could be used by end users to handle all sorts of custom formats hopefully well enough.

**Comment by nijel** [comment]
The regex based format will also not magically appear out of nothing :-). The issue hasn't progressed because nobody had time to move it forward, not because it's impossible to fix that.  Another reason might be that workaround described in dfc4bce3b67 works good enough for most of the users. The translate-toolkit itself is lacking contributors and most of changes in their recent releases were implemented by Weblate contributors. Over the time I've also became one of translate-toolkit maintainers because of that.

To move things forward you can develop the fix yourself or find somebody interested in implementing this. You can also pay me or my colleague to implement this, see https://weblate.org/support/.

**Comment by andrewleech** [comment]
Oh yes, sorry I didn't mean to put pressure on you, this was as much throwing out the idea to see what you thought. I maintain and contribute to far too many open source projects myself so understand the time and priorities challenges.

I will likely look at implementing it myself at some stage, thanks for the feedback!

**Comment by nijel** [comment]
Revisiting this again, I don't think regex is good idea here. It might work for parsing, but we need also serialize the file and that's something where regex is not helpful at all.

**Comment by andrewleech** [comment]
I had thought that the regex search match objects could be equally used to provide the location details for where the replacement needs to go, using like a regex text replacement.
